File: Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

bool luhn_check(long number);
int get_length(long number);

int main(void) {
    long number = get_long("Number: ");

    if (!luhn_check(number)) {
        printf("INVALID\n");
        return 0;
    }

    int len = get_length(number);

    long start_digits = number;
    while (start_digits >= 100) {
        start_digits /= 10;
    }

    if (len == 15 && (start_digits == 34 || start_digits == 37)) {
        printf("AMEX\n");
    } else if (len == 16 && (start_digits >= 51 && start_digits <= 55)) {
        printf("MASTERCARD\n");
    } else if ((len == 13 || len == 16) && (start_digits / 10 == 4)) {
        printf("VISA\n");
    } else {
        printf("INVALID\n");
    }

    return 0;
}

bool luhn_check(long number) {
    int total = 0;
    int pos = 0;

    while (number > 0) {
        int digit = number % 10;

        if (pos % 2 == 1) {
            digit *= 2;
            if (digit > 9) {
                digit -= 9;
            }
        }

        total += digit;
        number /= 10;
        pos++;
    }

    return (total % 10 == 0);
}

int get_length(long number) {
    int length = 0;
    while (number > 0) {
        number /= 10;
        length++;
    }
    return length;
}

File: Problem Set 6/sentimental-credit/credit.py
def main():
    number = input("Number: ")

    if not luhn_check(number):
        print("INVALID")
        return

    if len(number) == 15 and (number.startswith("34") or number.startswith("37")):
        print("AMEX")
    elif len(number) == 16 and number[:2] in {"51", "52", "53", "54", "55"}:
        print("MASTERCARD")
    elif len(number) in {13, 16} and number.startswith("4"):
        print("VISA")
    else:
        print("INVALID")

def luhn_check(number):
    total = 0
    reverse_digits = number[::-1]

    for i, digit in enumerate(reverse_digits):
        n = int(digit)
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9
        total += n

    return total % 10 == 0

if __name__ == "__main__":
    main()
